### DISCUSSION TITLE
Dudas sobre ejercicio filesystem

### DISCUSSION THREAD

**Discussion by foviedo** [dds-jv/foro#128] [Consultas de objetos] [answered]
Tengo un par de dudas respecto a este ejercicio:
Creo que está faltando un repo, encontré este https://github.com/dds-utn/file-system pero sólo tiene descripciones pero nada de código para poder testear ni los tests en sí que dice que tienen
Por otro lado, cómo es que el usuario me va a mandar qué quere leer? Con inicio y final del buffer? O sólo con cuánto quiere leer y yo me preocupo por mover el buffer bien? Cómo funciona eso del Consumer<Integer> callback? Qué es eso de C0, C1 y C2? Y lo de escribir bloques de 0x0, 0x10...? 
Saludos,
Facundo

**Comment by gastonprieto** [comment]
> Creo que está faltando un repo, encontré este https://github.com/dds-utn/file-system pero sólo tiene descripciones pero nada de código para poder testear ni los tests en sí que dice que tienen

En realidad esta perfecto, ni siquiera necesitas ese repositorio. En el documento tenes la interfaz de un componente para operar con el file system, no necesitas conocer la implementación. 
Y para testear vimos la clase pasada herramientas para testear, los mocks.

Es muy común que tengas ejercicios donde te dan una interfaz y solo tengas que entender como usarla. 

> Por otro lado, cómo es que el usuario me va a mandar qué quere leer? Con inicio y final del buffer? O sólo con cuánto quiere leer y yo me preocupo por mover el buffer bien?

Esto depende de como definas vos tu nuevo componente. Lo único que dice que tengas en cuenta es que sea fácil para el usuario? ¿Cúal decis que es la mas fácil?

> Cómo funciona eso del Consumer callback

Un consumer es el tipo de una lambda en este caso que recibe un entero. Tiene que ver con lo que esta en el apunte de patrones de comunicación

> Qué es eso de C0, C1 y C2

C0 = campo 0, C1 = campo 1, etc.

> de 0x0, 0x10

Son ejemplos de bytes a escribir, pero estan expresados en hexadecimal.

**Comment by gastonprieto** [comment]
@foviedo te sirvio? te quedo alguna duda?

**Comment by foviedo** [comment]
Me sirvió! Gracias. Pero me vienen otras dudas ahora:
-Nos gustaría que la interfaz nos expusiera al menos una forma de aplicarle a cada bloque una transformación, o de filtrar fácilmente bloque
No estaría entendiendo a qué se refiere
Y
-Nos gustaría que funcionara para archivos muy grandes, potencialmente infinitos.
Si bien entiendo que acá hay que usar stream, no me termina de cerrar de qué forma, hasta ahora sólo vimos stream para el map y el filter por eso y no veo bien cómo usar map y filter acá

**Comment by gastonprieto** [comment, marked as answer]
> Nos gustaría que la interfaz nos expusiera al menos una forma de aplicarle a cada bloque una transformación, o de filtrar fácilmente bloque

Fijate que las lecturas se hacen de a cachos (inicio-fin), lo mismo que la escritura, eso puede entenderse como un bloque.

> Nos gustaría que funcionara para archivos muy grandes, potencialmente infinitos. Si bien entiendo que acá hay que usar stream, no me termina de cerrar de qué forma, hasta ahora sólo vimos stream para el map y el filter por eso y no veo bien cómo usar map y filter acá

Esto lo seguimos por #34 como encararlo. Pero igualmente map/filter son operaciones de alto nivel que suelen aparecer en los streams, pero no son las unicas, lo mas importante es que es un objeto que va a generar valores (potencialmente infinitos) y se los va a pasar a la operación que le digan que tiene que ejecutar cuando hay un valor nuevo, por ejemplo

Pseudocódigo
```
stream = Stream.create(1,2,3)
stream.each(nro => console.log(nro))
```

`stream` representa los valores `1,2,3` y la operacion a ejecutar por cada uno de los elementos es `nro => console.log(nro)`

**Comment by gastonprieto** [comment]
@foviedo sirvio esto ? te quedo alguna duda mas ?

**Comment by foviedo** [comment]
Nada más, gracias!
